### PR TITLE
[dune] [doc] Support for building the reference manual with Dune.

### PR DESCRIPTION
**Pull request by ejgallego**
This is a reduced version of #8503 as to provide a way to build the
reference manual with Dune.

Dune 1.6 supports (experimentally) directories as targets, thus we
introduce a rule that will call `sphinx` to build the manual.

This only provides build, however generation of the `.install` file is
still incomplete, and it will be hopefully addressed in #8503.

### CODE REVIEW

File: doc/tools/coqrst/coqdoc/main.py
Comment (cpitclaudel) [inline doc/tools/coqrst/coqdoc/main.py]: You can pass a default to getenv instead of using `or ""`. But I'm not sure this is a good idea: do you really want to run the system-global coqdoc here?
This code was written with the assumption that callers would either set COQBIN or pass and explicit coqdoc_bin

Comment (ejgallego) [reply doc/tools/coqrst/coqdoc/main.py]: > You can pass a default to getenv instead of using `or ""`. But I'm not sure this is a good idea: do you really want to run the system-global coqdoc here?

Yes, we want to run with the `coqdoc` in `PATH` here; using `COQBIN` has a problem as it doesn't properly compose when building several packages together.

Should I update and remove the `or` then?

Comment (ejgallego) [reply doc/tools/coqrst/coqdoc/main.py]: Dune composes builds _à la Nix_ , that is to say, `PATH` is modified so you have the right scope. Starting in Dune 1.6 we can also scope private binaries, which helps with some problems we have with the test suite and `fake_ide`.

**Comment by ejgallego** [comment]
Build by this PR https://coq.gitlab.io/-/coq/-/jobs/130461993/artifacts/_build/default/doc/sphinx_build/html/index.html

Reading it now to verify there are no oddities.

**Comment by ejgallego** [comment]
> Reading it now to verify there are no oddities.

At a quick glance the build seems fine, thus this should be ready modulo the comments by Clément; IMHO as the refman will go on its own Debian/OPAM/Nix package having the possibility to use the installed `coqtop` does make sense to me.

File: doc/tools/coqrst/coqdoc/main.py
Comment (cpitclaudel) [reply doc/tools/coqrst/coqdoc/main.py]: Yes, changing the or would be good, then. It's too bad that this will make mistakes when running the script on its own trickier to spot, but I don't think that's an important enough issue to delay this PR.

Comment (ejgallego) [reply doc/tools/coqrst/coqdoc/main.py]: I see what you mean, on the other hand I'd dare to say running via path is a pretty common occurrence these days in sandboxed-style builds. For example I think that people on Nix should not have any other `coqtop` in scope.

Having to pass parameters becomes quickly a mess too if you context is large tho, and tends to have portability issues.

Comment (ejgallego) [reply doc/tools/coqrst/coqdoc/main.py]: Amended; I am still open as how to improve the general setup for people that don't like to use `PATH` in general; this is an important issue as we have two core dev setups [Nix and Dune] that rely on it, so we must try to accommodate both options in the best possible way.

Comment (cpitclaudel) [reply doc/tools/coqrst/coqdoc/main.py]: Thanks. I think the code is OK as is. The problem with picking up the wrong Coq version would only pop up if you ran the script standalone, not as part of the Makefile or through Dune. Do we have to worry about this? I think probably not. If we do want to worry, we could probably add another env var.

Comment (Zimmi48) [reply doc/tools/coqrst/coqdoc/main.py]: Indeed, I don't think that we should have to worry about this. Soon Dune will be the only supported way of building BTW.

Comment (cpitclaudel) [reply doc/tools/coqrst/coqdoc/main.py]: I meant that with this change, sphinx-build alone in the refman directory may yield surprising results. That's not a huge issue.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: This is not actually simplifying anything, don't use template when everything in it except for a trivial line gets overridden.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]

File: Makefile.dune
Comment (ejgallego) [reply Makefile.dune]: Done, thanks.

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: Well, to me it servers a very important function here which is documentation; also it is a seed for future refactoring so I'd be in favor of keeping it as it doesn't create an obvious problem.

File: Makefile.dune
Comment (ejgallego) [reply Makefile.dune]: I've also moved from the other PR the refactoring of the PHONY line.

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: What does it document?
As for future refactoring, don't anticipate it by making things more complicated.

Comment (ejgallego) [reply .gitlab-ci.yml]: It documents this target is using a Dune-based build.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: >It documents this target is using a Dune-based build.

Not worth it IMO. Please change.

Comment (Zimmi48) [reply .gitlab-ci.yml]: Let's not block this PR on this unrelated change. @ejgallego you can push it again once the benefits are clearer so that @SkySkimmer doesn't oppose.

Comment (ejgallego) [reply .gitlab-ci.yml]: I'm afraid we face a difference of opinion here. In particular I don't want to have N targets when we move from `make -f Makefile.dune` to `dune` itself; and doing a search for `dune-template` is important in my workflow.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: Then move everything except `make -f Makefile dune` from dune-template to build:dune since it's specialised for it.

Comment (ejgallego) [reply .gitlab-ci.yml]: `dune-template` is already in `master`, thus apart from quite deep nitpicking indeed the comment is unrelated to this PR.

Comment (ejgallego) [reply .gitlab-ci.yml]: Ok, let's see now.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: NB: this change means this artifact now has an expire key. (maybe make a note in the commit message)

**Review by SkySkimmer — APPROVED** [review]

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: Ok. I do prefer to expire by default, specially in ephemeral builds; but of course this could be overriden if our deployment strategy changes.

Comment (ejgallego) [reply .gitlab-ci.yml]: Maybe one month could be a more sensible outcome here tho.

**Comment by ejgallego** [comment]
Well, this has a problem that indeed I just realized:
- our default dune target is `dune build @install`,
- we now install the `coq-refman` package,
- thus the refman is built on the base target,

The `doc:refman:dune` target should not build anything, just upload the artifact.

There are different ways to solve this, maybe for now we should just guard the build of the refman alias somehow.

**Comment by SkySkimmer** [comment]
I don't know about "people" in general, but for CI at least we can be more specific in what we want built/installed.

File: Makefile.dune
Comment (Zimmi48) [inline Makefile.dune]: Well know the whole thing needs Dune >= 1.6, no?

**Review by Zimmi48 — COMMENTED** [review]
@ejgallego I'm waiting on your green light before merging this.

**Comment by rgrinberg** [comment]
> People should just not use the default target. WDYT?

Note that the default target can be changed to be whatever you want it to:

```
(alias
 (name default)
 (deps foo bar baz ..))
```

**Comment by Zimmi48** [comment]
Thanks @rgrinberg. @ejgallego Could you change the default to `coq.install`? This would avoid confusion.

**Comment by ejgallego** [comment]
I will tweak it, maybe I use a different method tho, like enabling the alias only iff an env var is set.

**Comment by ejgallego** [comment]
Ok, after some thinking I am convinced now that the right strategy for this PR is not to enable the install target for now. Anyways the resulting install file was unusable. Thus, the refman will be only be built if the target is specified explicitly.

File: dune-project
Comment (Zimmi48) [inline dune-project]: There are a few place in the doc that needs updating to reflect this: `Makefile.dune` and `dev/doc/build-system.dune.md`.

**Review by Zimmi48 — APPROVED** [review]
Last round of comments (totally optional ones):
What about updating `dev/doc/build-system.dune.md` to mention that there is experimental support for building the HTML version of the reference manual with Dune?

File: dev/doc/build-system.dune.md
Comment (ejgallego) [reply dev/doc/build-system.dune.md]: Umm, we need to be careful here, I'll had an "usually"

**Comment by ejgallego** [comment]
> @ejgallego I have just pushed to your branch a commit updating `default.nix` so that it provides Dune 1.6 and it remains possible to build Coq with Dune for Nix users.

Oh thanks!

> What about updating `dev/doc/build-system.dune.md` to mention that there is experimental support for building the HTML version of the reference manual with Dune?

Sure.

**Comment by ejgallego** [comment]
Thanks for all your help!

**Pull request merged**
